### PR TITLE
ReplicaSet ShardOperations

### PR DESCRIPTION
**Pull request by agourlay**
This PR adds an implementation of `ShardOperation` for `ReplicaSet`.

## Description

Following the proposal of https://github.com/qdrant/qdrant/issues/993, the `ReplicaSet` now manages several remote shards and becomes responsible for the shard query coordination.

There is an optimization scheme for the read-only operation in order to reduce the latency:
1. prefer a local shard
2. try a subset of the remote shards concurrently (the ratio could be fully configurable)
3. try remaining remote shards 

Regarding the update operations, the processing must be acknowledged by all shards to be well formed.
Failing peers are reported as `inactive` in the `replica_state`.

## Notes for the reviewers

I was not able to have a nice `update` path because of various struggles with `async` blocks and `Iterators`.
In short I wanted to somehow iterate over trait objects from `self.remotes.iter().chain(self.local.iter())` but I was unable to align the types :man_shrugging: 

I was planning to do the usual integration testing approach at the end.

### CODE REVIEW

File: lib/collection/src/shard/replica_set.rs [hunk 1]
```diff
@@ -75,52 +84,200 @@ impl ReplicaSet {
...
+                    None => Err(CollectionError::service_error(format!(
+                        "None of the replicas replied for Replica set {} on peer {}",
+                        self.shard_id, self.this_peer_id
+                    ))),
+                    Some(res) => Ok(res),
+                }
+            }
+            Err((peer_id, err)) => {
+                // report failing `peer_id`
+                self.set_active(&peer_id, false)?;
```
Comment (e-ivkov) [inline lib/collection/src/shard/replica_set.rs:233]: All peers need to sync the active state of replicas. So we can't use this shortcut here, instead submit a `SetShardReplicaState` operation to consensus

Comment (e-ivkov) [reply lib/collection/src/shard/replica_set.rs:233]: Btw if it is the only place that requires `mut` state, then we can safely remove RwLock

File: lib/collection/src/shard/replica_set.rs [hunk 2]
```diff
@@ -75,52 +84,200 @@ impl ReplicaSet {
...
+        }
+
+        // shortcut at first successful result
+        while let Some(result) = futures.next().await {
+            if let ok @ Ok(_) = result {
+                return ok;
+            }
+        }
+        // at this point `captured_error` must be defined by construction
+        captured_error.unwrap()
```
Comment (e-ivkov) [inline lib/collection/src/shard/replica_set.rs:163]:
```suggestion
        captured_error.expect("at this point `captured_error` must be defined by construction")
```

File: lib/collection/src/shard/replica_set.rs [hunk 3]
```diff
@@ -17,27 +22,28 @@ pub type IsActive = bool;
...
+/// Perform updates on all replicas and report error if there is at least one failure.
 pub struct ReplicaSet {
     shard_id: ShardId,
     this_peer_id: PeerId,
     local: Option<LocalShard>,
-    // TODO: Remote shard should be able to query several peers
-    remote: Option<RemoteShard>,
-    pub(crate) replica_state: HashMap<PeerId, IsActive>,
+    remotes: Vec<RemoteShard>,
+    pub(crate) replica_state: RwLock<HashMap<PeerId, IsActive>>,
```
Comment (e-ivkov) [inline lib/collection/src/shard/replica_set.rs:32]: Let's also add `SaveOnDisk<HashMap<PeerId, IsActive>>` as we discussed we need to retain these info. Of course this can also be done in a separate PR

Comment (agourlay) [reply lib/collection/src/shard/replica_set.rs:32]: Good point, I will do this in a follow up PR to keep things small :+1:

File: lib/collection/src/shard/replica_set.rs (hunk 1 above)
Comment (agourlay) [reply lib/collection/src/shard/replica_set.rs:233]: This is indeed the only place.

This means I need to submit the operation via the `Dispatcher` from the `ReplicaSet`?

However, the `Dispatcher` type lives in `Storage` and cannot be accessed from the `Collection` crate.
Is there another way?

Comment (agourlay) [reply lib/collection/src/shard/replica_set.rs]: The callback cannot be passed via the `ShardOperation` without messing up the trait interface.

Therefore I spent a long time trying to have the callback being saved at the level of the `ReplicaSet` struct but I am facing an issue.

I am getting the following error when trying to use the callback.

```
error[E0618]: expected function, found `(dyn std::ops::Fn(u64) -> (dyn futures::Future<Output = ()> + 'static) + std::marker::Send + std::marker::Sync + 'static)`
  --> lib/collection/src/shard/replica_set.rs:41:9
   |
41 |         (*self.notify_peer_failure_cb)(peer_id)
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^---------
   |         |
   |         call expression requires function

```

I need your Rust superpowers :pray:

Comment (e-ivkov) [reply lib/collection/src/shard/replica_set.rs]: I pushed the changes that should fix the problem. They are not full - I left some stubs for the tests that we need to fix before we can merge it. But now it type-checks!

Comment (e-ivkov) [reply lib/collection/src/shard/replica_set.rs]: It is just that under the hood async functions are built on top of pinned futures, I agree, that's not a very intuitive idea

Comment (agourlay) [reply lib/collection/src/shard/replica_set.rs]: I have cleaned up the stubs to please `clippy` and get green tests in https://github.com/qdrant/qdrant/pull/997/commits/323b6055f56dd31afca04dabf3d7fdf2e31104bb
I also introduced a dummy callback to restore the collection's tests.
WYDYT?

File: lib/storage/src/content_manager/toc.rs [hunk 4]
```diff
@@ -308,6 +309,27 @@ impl TableOfContent {
...
 
+    fn on_peer_failure_callback(&self) -> replica_set::OnPeerFailure {
+        let proposal_sender = self.consensus_proposal_sender.clone();
+        Box::new(move |peer_id| {
+            let proposal_sender = proposal_sender.clone();
+            Box::new(async move {
+                proposal_sender
+                    .send(ConsensusOperations::CollectionMeta(
+                        CollectionMetaOperations::SetShardReplicaState(SetShardReplicaState {
+                            // TODO: fill in actual collection name and shard_id
```
Comment (e-ivkov) [inline lib/storage/src/content_manager/toc.rs:320]: Can you fix this `todo` as well? I think it is in scope of this PR.

File: lib/collection/src/shard/replica_set.rs (hunk 1 above)
Comment (e-ivkov) [reply lib/collection/src/shard/replica_set.rs]: I think there is still 1 minor todo left to fix, but otherwise looks good!

File: lib/storage/src/content_manager/toc.rs (hunk 4 above)
Comment (agourlay) [reply lib/storage/src/content_manager/toc.rs:320]: Already on it :+1: 

It seems `ShardId` needs to be passed through the callback as well.

Comment (e-ivkov) [reply lib/storage/src/content_manager/toc.rs:320]: Yes, shouldn't be a problem I think

**Review by e-ivkov — APPROVED** [review]
